encoding/icicle/device_setup.go

```diff
@@ -0,0 +1,123 @@
+//go:build icicle
+
+package icicle
+
+import (
+	"fmt"
+	"log/slog"
+	"sync"
+
+	"github.com/consensys/gnark-crypto/ecc/bn254"
+	"github.com/ingonyama-zk/icicle/v3/wrappers/golang/core"
+	icicle_bn254 "github.com/ingonyama-zk/icicle/v3/wrappers/golang/curves/bn254"
+	icicle_runtime "github.com/ingonyama-zk/icicle/v3/wrappers/golang/runtime"
+)
+
+// IcicleDevice wraps the core device setup and configurations
+type IcicleDevice struct {
+	Device         icicle_runtime.Device
+	NttCfg         core.NTTConfig[[icicle_bn254.SCALAR_LIMBS]uint32]
+	MsmCfg         core.MSMConfig
+	FlatFFTPointsT []icicle_bn254.Affine
+	SRSG1Icicle    []icicle_bn254.Affine
+}
+
+// IcicleDeviceConfig holds configuration options for device setup
+type IcicleDeviceConfig struct {
+	EnableGPU bool
+	NTTSize   uint8
+	// MSM setup parameters (optional)
+	FFTPointsT [][]bn254.G1Affine
+	SRSG1      []bn254.G1Affine
+}
+
+// NewIcicleDevice creates and initializes a new IcicleDevice
+func NewIcicleDevice(config IcicleDeviceConfig) (*IcicleDevice, error) {
+	icicle_runtime.LoadBackendFromEnvOrDefault()
+
+	device := setupDevice(config.EnableGPU)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	var (
+		nttCfg         core.NTTConfig[[icicle_bn254.SCALAR_LIMBS]uint32]
+		msmCfg         core.MSMConfig
+		flatFftPointsT []icicle_bn254.Affine
+		srsG1Icicle    []icicle_bn254.Affine
+		setupErr       error
+		icicleErr      icicle_runtime.EIcicleError
+	)
+
+	// Setup NTT and optionally MSM on device
+	icicle_runtime.RunOnDevice(&device, func(args ...any) {
+		defer wg.Done()
+
+		// Setup NTT
+		nttCfg, icicleErr = SetupNTT(config.NTTSize)
+		if icicleErr != icicle_runtime.Success {
+			setupErr = fmt.Errorf("could not setup NTT")
+			return
```

Comment:
cannot return because I don't think that RunOnDevice func will let me. Looks like this:

```
func RunOnDevice(device *Device, funcToRun func(args ...any), args ...any) {
    go func(deviceToRunOn *Device) {
        defer runtime.UnlockOSThread()
        runtime.LockOSThread()
        originalDevice, _ := GetActiveDevice()
        SetDevice(deviceToRunOn)
        funcToRun(args...)
        SetDevice(originalDevice)
    }(device)
}
```

Reply:
ah my bad I completely missed the function signature.
I see so the code passed to RunOnDevice is run on the gpu...? Did you test that your shared pointer for setupErr is working then? This would mean the gpu is able to go write in the cpu's RAM. Does this only work on mac where cpu/gpu share memory? Or is this something that always works?

Reply:
No, RunOnDevice locks the OS thread so it makes all of that code run on the same OS thread. It's because with go routines you can call `SetDevice` to set the GPU device and that can run on thread 1 but the go routine can end up switching to thread 2 in the middle which means that it will use the default device (CPU).